src/topology/builder.rs

```diff
@@ -118,27 +120,37 @@ pub fn build_pieces(
 
         let typetag = source.source_type();
 
-        let server = match source.build(&name, &config.global, tx) {
+        let (global_shutdown_begun_trigger, global_shutdown_begun_tripwire) = Tripwire::new();
+        let (local_shutdown_complete_trigger, local_shutdown_complete_tripwire) = Tripwire::new();
+        let shutdown = ShutdownSignals {
+            begin_shutdown: global_shutdown_begun_tripwire.clone(),
+            shutdown_complete: local_shutdown_complete_trigger,
+        };
+
+        let server = match source.build(&name, &config.global, shutdown, tx) {
             Err(error) => {
                 errors.push(format!("Source \"{}\": {}", name, error));
                 continue;
             }
             Ok(server) => server,
         };
 
-        let (trigger, tripwire) = Tripwire::new();
-
         let (output, control) = Fanout::new();
         let pump = rx.forward(output).map(|_| ());
         let pump = Task::new(&name, &typetag, pump);
 
-        let server = server.select(tripwire.clone()).map(|_| ()).map_err(|_| ());
+        // TODO: Remove this line once all Sources do this internally.
+        let server = server
+            .select(global_shutdown_begun_tripwire)
```

Comment:
I think this select is related to this https://github.com/timberio/vector/pull/1917/files#r383582914

We need someway to force cancel the task if its taking too long. One trick is to wrap the source future with a select future that will drop the source future when it receives the force cancel command.

Reply:
Thanks, digging into this a bit more I think I understand how this works better now.  Please check out the addition of the 'shutdown_force_triggers' I added in the new version which I believe better handles this now.